underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java

```diff
@@ -11,478 +11,187 @@
-            if (!file.delete()) {
-              LOG.error("Failed to delete temporary file @ {}", file.getPath());
-            }
-          }
-          throw new IOException("Fail to upload part " + request.getPartNumber()
-              + " to " + request.getKey(), lastException);
-        });
-    mTagFutures.add(futureTag);
-    LOG.debug("Submit upload part request. key={}, partNum={}, file={}, fileSize={}, lastPart={}.",
-        mKey, request.getPartNumber(), file.getPath(), file.length(), request.isLastPart());
+  @Override
+  protected boolean isMultiPartUploadInitialized() {
+    return mUploadId != null;
   }
 
-  /**
-   * Waits for the submitted upload tasks to finish.
-   */
-  protected void waitForAllPartsUpload() throws IOException {
-    int beforeSize = mTags.size();
+  @Override
+  protected void createEmptyObject(String key) throws IOException {
     try {
-      for (ListenableFuture<PartETag> future : mTagFutures) {
-        mTags.add(future.get());
-      }
-    } catch (ExecutionException e) {
-      // No recover ways so that we need to cancel all the upload tasks
-      // and abort the multipart upload
-      Futures.allAsList(mTagFutures).cancel(true);
-      abortMultiPartUpload();
-      throw new IOException("Part upload failed in multipart upload with "
-          + "id '" + mUploadId + "' to " + mKey, e);
-    } catch (InterruptedException e) {
-      LOG.warn("Interrupted object upload.", e);
-      Futures.allAsList(mTagFutures).cancel(true);
-      abortMultiPartUpload();
-      Thread.currentThread().interrupt();
-    }
-    mTagFutures = new ArrayList<>();
-    if (mTags.size() != beforeSize) {
-      LOG.debug("Uploaded {} partitions of id '{}' to {}.", mTags.size(), mUploadId, mKey);
+      ObjectMetadata meta = new ObjectMetadata();
+      meta.setContentLength(0);
+      meta.setContentType(Mimetypes.MIMETYPE_OCTET_STREAM);
+      getClient().putObject(
+          new PutObjectRequest(mBucketName, key, new ByteArrayInputStream(new byte[0]), meta));
+    } catch (SdkClientException e) {
+      throw new IOException(e);
     }
   }
 
-  protected void completeMultiPartUpload() throws IOException {
-    completeMultiPartUpload(getClient(), mUploadId);
-  }
-
-  /**
-   * Completes multipart upload.
-   */
-  protected void completeMultiPartUpload(AmazonS3 s3Client, String uploadId) throws IOException {
-    AmazonClientException lastException;
-    CompleteMultipartUploadRequest completeRequest = new CompleteMultipartUploadRequest(mBucketName,
-        mKey, uploadId, mTags);
-    do {
-      try {
-        s3Client.completeMultipartUpload(completeRequest);
-        LOG.debug("Completed multipart upload for key {} and id '{}' with {} partitions.",
-            mKey, uploadId, mTags.size());
-        return;
-      } catch (AmazonClientException e) {
-        lastException = e;
+  @Override
+  protected void putObject(String key, File file, String md5) throws IOException {
```

Comment:
Why this method? where it's used?

Reply:
If the uploaded file size is smaller than the partition size, putObject will be used.